File: raw/github/paypal/paypal-ios-sdk-demo-app/snapshots/2026-08-16-047a50e/files/PayPalDemo/ViewModels/PayPalViewModel.swift
import SwiftUI
import PayPalWebPayments
import CorePayments
import FraudProtection

class PayPalViewModel: ObservableObject {

    private var payPalWebCheckoutClient: PayPalWebCheckoutClient?
    private var payPalDataCollector: PayPalDataCollector?

    func startCheckout(
        amount: Double,
        intent: Intent
    ) async throws -> String? {
        async let config = try await DemoMerchantAPI.shared.getCoreConfig()

        let order = try await DemoMerchantAPI.shared.createOrder(
            orderParams: CreateOrderParams(
                applicationContext: nil,
                intent: intent.rawValue,
                purchaseUnits: [PurchaseUnit(amount: Amount(currencyCode: "USD", value: "\(amount)"))]
            )
        )
        print("✅ Order created with orderID: \(order.id) with status: \(order.status)")

        payPalWebCheckoutClient = try await PayPalWebCheckoutClient(config: config)
        guard let payPalWebCheckoutClient = payPalWebCheckoutClient else {
            throw NSError(domain: "PayPalWebCheckoutClientError", code: -1, userInfo: [NSLocalizedDescriptionKey: "PayPalWebCheckout client could not be initialized."])
        }

        payPalDataCollector = try await PayPalDataCollector(config: config)
        let payPalClientMetadataID = payPalDataCollector?.collectDeviceData()

        do {
            let paypalCheckoutResult = try await payPalWebCheckoutClient.start(
                request: PayPalWebCheckoutRequest(
                    orderID: order.id,
                    fundingSource: .paypal)
            )
            print("✅ Order approved with orderID: \(paypalCheckoutResult.orderID) and PayerID: \(paypalCheckoutResult.payerID)")

            let completedOrder = try await DemoMerchantAPI.shared.completeOrder(
                orderID: order.id,
                payPalClientMetadataID: payPalClientMetadataID,
                intent: intent
            )
            print("✅ Capture returned with orderID: \(completedOrder.id) with status: \(completedOrder.status) ")
            return completedOrder.id
        } catch {
            // Convenience method if separate handling of cancel error is needed
            if PayPalError.isCheckoutCanceled(error) {
                return nil
            } else {
                throw error
            }
        }
    }
}

File: raw/github/paypal/paypal-android-sdk-demo-app/snapshots/2026-08-16-d1137d5/files/app/src/main/java/com/firstapp/paypaldemo/paypalcheckout/PayPalViewModel.kt
package com.firstapp.paypaldemo.paypalcheckout

import android.content.Context
import android.content.Intent
import androidx.activity.ComponentActivity
import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.firstapp.paypaldemo.Constants.CLIENT_ID
import com.firstapp.paypaldemo.Constants.DEEP_LINK_URL_SCHEME
import com.firstapp.paypaldemo.Constants.SHOPPING_CART_ITEMS
import com.firstapp.paypaldemo.main.CartUiState
import com.firstapp.paypaldemo.main.CheckoutState
import com.firstapp.paypaldemo.service.Amount
import com.firstapp.paypaldemo.service.DemoMerchantAPI
import com.firstapp.paypaldemo.service.PurchaseUnit
import com.paypal.android.corepayments.CoreConfig
import com.paypal.android.paypalwebpayments.PayPalPresentAuthChallengeResult
import com.paypal.android.paypalwebpayments.PayPalWebCheckoutClient
import com.paypal.android.paypalwebpayments.PayPalWebCheckoutFinishStartResult
import com.paypal.android.paypalwebpayments.PayPalWebCheckoutFundingSource
import com.paypal.android.paypalwebpayments.PayPalWebCheckoutRequest
import dagger.hilt.android.lifecycle.HiltViewModel
import dagger.hilt.android.qualifiers.ApplicationContext
import jakarta.inject.Inject
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

@HiltViewModel
class PayPalViewModel @Inject constructor(
    private val savedStateHandle: SavedStateHandle,
    @ApplicationContext context: Context
) : ViewModel() {

    private val coreConfig = CoreConfig(CLIENT_ID)
    private val payPalClient =
        PayPalWebCheckoutClient(context, coreConfig, DEEP_LINK_URL_SCHEME)

    private var _uiState = MutableStateFlow(defaultCartUiState)
    val uiState: StateFlow<CartUiState> = _uiState.asStateFlow()

    private var checkoutState
        get() = _uiState.value.checkoutState
        set(value) = _uiState.update { prevState -> prevState.copy(checkoutState = value) }

    init {
        savedStateHandle.get<String>(INSTANCE_STATE_KEY)?.let {
            payPalClient.restore(it)
        }
    }

    /**
     * Launches the PayPal web checkout flow via Braintree browser switch library
     * Braintree browser switch library is a wrapper for Chrome Custom Tab.
     * onSuccess means the user was successfully sent to the browser.
     * The final 'finish' must still happen in handleOnNewIntent => finishPayPalCheckout.
     */
    fun startPayPalCheckout(activity: ComponentActivity) {
        checkoutState = CheckoutState.OrderCreateInProgress("Starting PayPal Checkout")
        viewModelScope.launch {
            try {
                val items = _uiState.value.items
                val purchaseUnits = items.map { item ->
                    PurchaseUnit(amount = Amount(currencyCode = "USD", item.amount.toString()))
                }
                val order =
                    DemoMerchantAPI.createOrder(intent = "CAPTURE", purchaseUnits = purchaseUnits)
                println("✅ Created order ${order.id}, status: ${order.status}")

                val fundingSource = PayPalWebCheckoutFundingSource.PAYPAL
                val request =
                    PayPalWebCheckoutRequest(orderId = order.id, fundingSource = fundingSource)
                payPalClient.start(activity, request) { result ->
                    checkoutState = when (result) {
                        is PayPalPresentAuthChallengeResult.Success -> {
                            // persist instance state in case we need it after process kill
                            savedStateHandle[INSTANCE_STATE_KEY] = payPalClient.instanceState
                            CheckoutState.StartPayPalInProgress("Starting PayPal Checkout")
                        }

                        is PayPalPresentAuthChallengeResult.Failure ->
                            CheckoutState.Error(result.error.toString())
                    }
                }
            } catch (e: Exception) {
                val errorMessage = "❌ Failed to create order on merchant server: ${e.message}"
                checkoutState = CheckoutState.Error(errorMessage)
            }
        }
    } // startPayPalCheckout

    /**
     * Called after the user returns from the Chrome Custom Tab to finish the checkout.
     */
    fun finishPayPalCheckout(intent: Intent?) {
        val result = intent?.let { payPalClient.finishStart(it) }
        when (result) {
            is PayPalWebCheckoutFinishStartResult.Success -> {
                val orderId = result.orderId
                if (orderId == null) {
                    checkoutState =
                        CheckoutState.Error("received success but PayPal returned a null orderId")
                } else {
                    completeOrder(orderId)
                }
            }

            is PayPalWebCheckoutFinishStartResult.Failure ->
                checkoutState = CheckoutState.Error(result.error.toString())

            is PayPalWebCheckoutFinishStartResult.Canceled ->
                checkoutState = CheckoutState.Error("Checkout canceled by user.")

            is PayPalWebCheckoutFinishStartResult.NoResult, null -> {
                // Control has been passed to Chrome Custom Tab. The user's intent cannot be
                // determined by the SDK. By returning the UI to an idle state, we can give users
                // the opportunity to relaunch the flow e.g. if they accidentally closed
                // the Chrome Custom Tab and need to re-launch it
                _uiState.update { currentState ->
                    currentState.copy(
                        checkoutState = CheckoutState.Idle,
                        didInitiateCheckout = true
                    )
                }
            }
        }

        if (result != null) {
            // clear instance state to make sure we don't restore multiple times
            savedStateHandle.remove<String>(INSTANCE_STATE_KEY)
        }
    }

    private fun completeOrder(orderId: String) {
        viewModelScope.launch {
            val finalOrder = DemoMerchantAPI.completeOrder(orderId, "CAPTURE")
            println("✅ captured order: ${finalOrder.id}")
            checkoutState = CheckoutState.OrderComplete(finalOrder.id)
        }
    }

    companion object {
        const val INSTANCE_STATE_KEY = "paypal_client_instance_state"

        private val defaultCartUiState by lazy {
            val items = SHOPPING_CART_ITEMS
            val totalAmount = items.sumOf { it.amount }
            CartUiState(
                items = items,
                totalAmount = totalAmount,
                checkoutState = CheckoutState.Idle,
                didInitiateCheckout = false
            )
        }
    }
}
